### PR TITLE
Wrap `url-cpp`

### PR DESCRIPTION
**Pull request by dlecocq**
Provides a Cython extension that wraps `url-cpp` for great performance improvement. The speedup of course depends on the task and the input, but in the range or 20-50x for parsing and normalization tasks.

In a test of 7.75M URL pairs taken from the wild (from 1M pages sampled from customer campaigns), `url-cpp` and the old `url-py` bindings disagreed on the interpretation of 122 URLs. The majority of these cases are described in new tests and are either 1) acceptably rare, 2) bugs in the old `url-py`, or 3) acceptable new rulings on edge cases previously not tested. They are described in additional tests as well:

- The `port` must be an integer, and in the range [0-65535]. An exception is now thrown otherwise.
- The old `url-py` was apparently interpreting `https://t…/` (with ellipsis as a single character) as `https://t.../` (three periods) instead of our new interpretation which considers that a single non-basic codepoint -- `https://xn--t-9hn/`
- Relative URLs (without a host) now may have `punycode` and `unpunycode` called on them -- I'm not sure why we previously chose to throw an exception in that case
- Always include a trailing slash for empty paths (`URL.parse('http://foo.com').utf8 == 'http://foo.com/'`)
- Relative URL evaluation always 1) provides a trailing slash, 2) squashes excess `..`'s. This is in-line with the interpretation `abspath()` takes:

```python
URL.parse('http://base/').relative('../../../').utf8
# Old:
'http://base/../../../'
# New:
'http://base/'

URL.parse('http://base/').relative('../../../').abspath().utf8
# Old:
'http://base'
# New:
'http://base/'
```

@neilmb @b4hand @lindseyreno @vadim-moz

### CODE REVIEW

**Comment by dlecocq** [comment]
Based off of #40

File: setup.py [hunk 1]
```diff
@@ -22,10 +22,35 @@
 # WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
 from distutils.core import setup
+from distutils.extension import Extension
+
+ext_files = [
+	'url/url-cpp/src/url.cpp',
```
Comment (b4hand) [inline setup.py]: Is this using tabs? The indentation looks extra wide.

File: test.py [hunk 2]
```diff
@@ -177,14 +191,6 @@ def test(bad, good):
         yield test, bad, good
 
 
-def test_strict_unicode_escape():
```
Comment (b4hand) [inline test.py]: Why did this test disappear?

File: url/url.pyx [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+    property userinfo:
+        def __get__(self):
+            return self.ptr.userinfo()
+        def __set__(self, s):
+            self.ptr.setUserinfo(s)
+
+    def copy(self):
+        '''Return a new instance of an identical URL.'''
+        new = URL('')
+        new.ptr.setScheme(self.ptr.scheme())
```
Comment (b4hand) [inline url/url.pyx]: Seems like maybe you should add a copy constructor to `url-cpp` instead of implementing this in Cython.

File: url/url.pyx [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+            _other = self.parse(other, encoding)
+        else:
+            _other = self.parse(other.utf8)
+
+        _self = self.parse(self.utf8, 'utf-8')
+        _self.canonical().defrag().abspath().escape().punycode().remove_default_port()
+        _other.canonical().defrag().abspath().escape().punycode().remove_default_port()
+
+        return (
+            _self.scheme    == _other.scheme    and
```
Comment (b4hand) [inline url/url.pyx]: Again, this is an argument for implementing `operator==` in `url-cpp`.

File: url/url.pyx [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+            _self.host      == _other.host      and
+            _self.path      == _other.path      and
+            _self.params    == _other.params    and
+            _self.query     == _other.query     and
+            _self.port      == _other.port
```
Comment (b4hand) [inline url/url.pyx]: Why is `port` last in this check rather than after the `host` to match order of appearance?

File: url/url.pyx [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+        return self
+
+    def defrag(self):
+        '''Remove the fragment from this url'''
+        self.ptr.defrag()
+        return self
+
+    def deparam(self, params):
+        '''Strip any of the provided parameters out of the url'''
+        lowered = unordered_set[string](p.lower() for p in params)
```
Comment (b4hand) [inline url/url.pyx]: Doesn't lowercasing the params here mean you won't be able to strip a parameter named `CAPITAL=1`?

File: url/url.pyx [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+        self.ptr.abspath()
+        return self
+
+    def sanitize(self):
+        '''A shortcut to abspath and escape'''
+        self.ptr.abspath().escape(False)
+        return self
+
+    def remove_default_port(self):
+        '''If a port is provided an is the default, remove it.'''
```
Comment (b4hand) [inline url/url.pyx]: s/provided an is the default/provided and is the default/

File: setup.py (hunk 1 above)
Comment (dlecocq) [inline setup.py]: D'oh.

File: test.py (hunk 2 above)
Comment (dlecocq) [inline test.py]: `URL`s can no longer be constructed component-wise like this, and are converted from unicode to UTF-8 at `parse` time.

File: url/url.pyx (hunk 3 above)
Comment (dlecocq) [inline url/url.pyx]: Done. Rather, accomplished with a construction and then an `assign`.

File: url/url.pyx (hunk 4 above)
Comment (dlecocq) [inline url/url.pyx]: Doen

File: url/url.pyx (hunk 5 above)
Comment (dlecocq) [inline url/url.pyx]: `port`-ed (nyuk nyuk) to `url-cpp`

File: url/url.pyx (hunk 6 above)
Comment (dlecocq) [inline url/url.pyx]: `deparam` is case-insensitive.

File: url/url.pyx (hunk 7 above)
Comment (dlecocq) [inline url/url.pyx]: Done

**Comment by dlecocq** [comment]
Reupped

File: url/url.pyx (hunk 6 above)
Comment (b4hand) [inline url/url.pyx]: I thought we discussed making `deparam` case sensitive?

Comment (dlecocq) [inline url/url.pyx]: That's not ringing a bell. I know we talked about casing in the context of nameprep casemapping stuff, but my memory was that we were content with continuing the use of case-insensitive deparams.

**Comment by b4hand** [comment]
LGTM
